pkg/quota/admission/podlimitrequest/admission.go

```diff
@@ -0,0 +1,152 @@
+	glog.V(5).Infof("PodLimitRequest admission controller is loaded")
+	parsed, err := ReadConfig(config)
+	glog.V(5).Infof("PodLimitRequest admission controller got config: %v\nerror: %v", parsed, err)
+	return &podLimitRequest{
+		Handler: admission.NewHandler(admission.Create),
+		Config:  parsed,
+	}, err
+}
+
+func ReadConfig(configFile io.Reader) (api.PodLimitRequestConfig, error) {
+	glog.V(5).Infof("PodLimitRequest about to read config:\n%v", configFile)
+	config := api.PodLimitRequestConfig{}
+	buffer := new(bytes.Buffer)
+	if _, err := buffer.ReadFrom(configFile); err != nil {
+		return config, err
+	}
+	err := yaml.Unmarshal(buffer.Bytes(), &config)
+	glog.V(5).Infof("PodLimitRequest config:\n%v\nerror: %v", config, err)
+	return config, err
+}
+
+func Validate(config api.PodLimitRequestConfig) error {
+	if config.Enabled {
+		if config.LimitCPUToMemoryRatio == 0.0 && config.CPURequestToLimitRatio == 0.0 && config.MemoryRequestToLimitRatio == 0.0 {
+			return fmt.Errorf("PodLimitRequest plugin enabled but no ratios specified")
+		}
+		if config.LimitCPUToMemoryRatio < 0.0 {
+			return fmt.Errorf("LimitCPUToMemoryRatio must be positive")
+		}
+		if config.CPURequestToLimitRatio < 0.0 || config.CPURequestToLimitRatio > 1.0 {
+			return fmt.Errorf("CPURequestToLimitRatio must be between 0.0 and 1.0")
+		}
+		if config.MemoryRequestToLimitRatio < 0.0 || config.MemoryRequestToLimitRatio > 1.0 {
+			return fmt.Errorf("MemoryRequestToLimitRatio must be between 0.0 and 1.0")
+		}
+	}
+	return nil
+}
+func (a *podLimitRequest) Validate() error {
+	if err := Validate(a.Config); err != nil {
+		return err
+	}
+	if a.ProjectCache == nil {
+		return fmt.Errorf("PodLimitRequest did not get a project cache")
+	}
+	return nil
+}
+
+func (a *podLimitRequest) Admit(attr admission.Attributes) error {
+	glog.V(8).Infof("PodLimitRequest admission controller is invoked")
+	if !a.Config.Enabled {
+		return nil
+	}
+	if attr.GetResource() != kapi.Resource("pods") || attr.GetSubresource() != "" {
+		glog.V(8).Infof("PodLimitRequest: not a pod, ignoring")
+		return nil
+	}
+	pod := attr.GetObject().(*kapi.Pod)
+	glog.V(5).Infof("PodLimitRequest is looking at creating pod %s in project %s", pod.Name, attr.GetNamespace())
+
+	// allow annotations on project to override
+	if ns, err := a.ProjectCache.GetNamespace(attr.GetNamespace()); err != nil {
+		glog.Warningf("PodLimitRequest got an error retrieving namespace: %v", err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	} else {
+		projectEnabledPlugin, exists := ns.Annotations[podLimitRequestAnnotation]
+		if exists && projectEnabledPlugin != "true" {
+			glog.V(5).Infof("PodLimitRequest is disabled for project %s", attr.GetNamespace())
+			return nil // disabled for this project, do nothing
+		}
+	}
+
+	// This relies on LimitRanger already having applied limits from the project.
+	// It will then adjust values as configured; not going to re-validate against LimitRanger.
```

Comment:
@derekwaynecarr going to run before LimitRanger so it validates. Confusing as it would be to an end user if their admin managed to mess this up, it seems like the right thing to do.

Implementation detail: I'm thinking I'll pick up any defaults, then run these ratio overrides, then pass off to LimitRanger. Of course the logic for merging in default limits is in LimitRanger (k8s.io/kubernetes/plugin/pkg/admission/limitranger/admission.go). It would be convenient for this if defaultContainerResourceRequirements and mergePodResourceRequirements were exported. Which approach seems best to you:

1. Capitalize these and use them. Carry the patch until we can get that change upstream.
2. Just use PodLimitFunc which already invokes these, and ignore any errors returned.
3. Copy those methods' source over to my plugin.